### ISSUE TITLE
Bug: Validator client is ahead of current epoch

### ISSUE THREAD

**Issue by djoyahoy** [sigp/lighthouse#2280]
## Description

I have one beacon node and one validator running on separate pods in the same kubernetes namespace. They are running against the pyrmont test network. The validator is still submitting attestations correctly. However, the validator is intermittently reporting:

```
Mar 23 16:09:43.005 ERRO Failed to download proposer duties      err: All endpoints failed https://<URL> => RequestFailed(ServerMessage(ErrorMessage { code: 400, message: "BAD_REQUEST: request epoch 28164 is ahead of the current epoch 28163", stacktraces: [] })), service: duties
Mar 23 16:09:43.005 ERRO Failed to download attester duties      err: FailedToDownloadAttesters("All endpoints failed https://<URL> => RequestFailed(ServerMessage(ErrorMessage { code: 400, message: \"BAD_REQUEST: request epoch 28165 is more than one epoch past the current epoch 28163\", stacktraces: [] }))"), request_epoch: 28165, current_epoch: 28164, service: duties
```

The client and the beacon containers report the same time and are running on the same node so I don't suspect there is any time drift.

## Version

Lighthouse v1.2.1-b34a79d
rustc 1.50.0

## Present Behaviour

See description.

## Expected Behaviour

This error should not be reported.

## Steps to resolve

N/A

**Comment by michaelsproul** [comment]
I think this must be a consequence of #2243 where we removed a 100ms delay from the duties service. For now downgrading to v1.2.0 should be safe.

To resolve the issue I think we have a few options:

- Add the delay back in. I'm not fond of this option because it punishes nodes without clock sync issues by forcing them to always wait the delay. We could make the delay 0 by default and configurable, but that just feels like pushing the burden onto the end user.
- Modify the proposer and attester duties API endpoints on the beacon node so that they're capable of responding to queries for the current_epoch +1/+2. If the BN thinks it's in the last slot of the epoch, and it already has a (canonical) block for that slot, then it knows what the duties are for the next epoch, and will already have computed them via the state advance timer. If we extend the state advance timer to handle skip slots, then it will always know the duties for the next epoch near the end of the current epoch. This would open up the option of polling _before_ the start of the relevant slot.
- In addition to modifying the BN duties endpoints, we could get the VC to subscribe to the head events as [suggested by the API docs](https://ethereum.github.io/eth2.0-APIs/#/Validator/getProposerDuties), and use that to determine when the duties could have changed. It's push-based! No more polling!

CC @paulhauner, would like your input on the above

**Comment by paulhauner** [comment]
Hmm, interesting. AFAICT there's necessarily *some* clock drift happening here, but it could be very slight. I think we could do better to tolerate this drift, though.

> Add the delay back in. I'm not fond of this option...

I agree.

> Modify the proposer and attester duties API endpoints on the beacon node so that they're capable of responding to queries for the current_epoch +1/+2.

I also prefer this style of approach. My initial thoughts are to add a [`MAXIMUM_GOSSIP_CLOCK_DISPARITY`](https://github.com/sigp/lighthouse/blob/b34a79dc0b02e04441ba01fd0f304d1e203d877d/beacon_node/beacon_chain/src/beacon_chain.rs#L99-L102) tolerance to these endpoints. In effect, the BN would permit users to poll for duties up to 500ms before they become relevant. Some advantages of this approach:

- We already permit this tolerance on the gossip network. This means that we're not really changing anything when it comes to proposer or current-epoch attester duties; P2P users could already force us to learn these duties early. P2P couldn't trigger next-epoch attester duties, but I don't think it's a significant change.
- There are already some assumptions about the VC being within the clock disparity of the BN. We verify things from the API practically the same as things from the gossip network; we'll still publish blocks/attestations if they're early but within that tolerance.

> In addition to modifying the BN duties endpoints, we could get the VC to subscribe to the head events

This would be nice, but I'm not sure any of the present events would solve this issue. There's no event to say "a new epoch has started and therefore caused a change in the proposer/attester duties".

**Comment by paulhauner** [comment]
Resolved in #2283
